pages/blog/_posts/2020-01-28-Thanos-infinite-metrics.md

```diff
@@ -0,0 +1,97 @@
+
+### Thanos Querier
+Querier implements the Prometheus Query API and understands PromQL. It then sends the query using the aforementioned
+Store API to all known Thanos Stores (discovered using service discovery) and awaits the metric information from the
+stores, be it directly from Prometheus via the Sidecar or metrics stored in S3 Object storage via Thanos Store.
+
+### Thanos Compactor
+It does not make a lot of sense to keep old metrics that are scraped every 15 or 30 seconds forever. At one point those 
+metrics would no longer be useful to make sense of your metrics. This is where
+the Thanos Compactor comes in. It creates aggregates of old metrics based on rules. It will for example
+aggregate metrics that are older than 30 days into 5 minute chunks. This saves resources and still gives you
+almost the same accuracy when looking at longer timespans. After those metrics have been aggregated, they are
+written back into the S3 bucket and the metadata gets updated.
+
+### Thanos Ruler
+The Ruler component is the Thanos equivalent of Recording Rules. It can look at all Store APIs and generate new metrics
+according to the Recording Rules fed into the Ruler component. However since this rule processing is not done against a 
+local datastore, it is possible that these new metric datapoints will not always be generated as it relies on a reliable
+data source to do this in the required intervals.
+
+## Architecture in a Cluster
+
+![Thanos Architecture](/img/blog/thanos/thanos_architecture.png)
+
+As we can see, there are quite a few things going on in this architectural view of the system, but on the other hand it
+is fairly simple to understand as the component are nicely decoupled from each other.
+
+An interesting thing here is that the drawing has multiple Prometheus instances with multiple Sidecars. Thanos actually
+allows for deduplication of Timeseries Data. The data uploaded from the sidecars have information from which prometheus
+instance the metrics are generated and adds that. The Querier can then deduplicate this data so that the metrics shown
+in Grafana are consistent and do not come sometimes from Prometheus A and sometimes from Prometheus B.
+
+This design also allows for an interesting other Use Case: Querying multiple Clusters. As
+long as the storage location used by the Sidecars for uploading the metrics is identical, the Time Series data is
+available to the Thanos Store and therefor the Querier and Grafana.
+That even allows to run Grafana, Querier and Store in a completely different part of the world if need be.
+
+## Is it worth it?
+
+### Pros
+- Highly Available Prometheus
+- Increased Reliability (Decoupled query component)
+- Infinite Metric Storage
+- Query Multiple clusters from a single point
+- Easy to Scale
+
+### Cons
+- More complex architecture
+- Increased Resource usage
+
+## Summary
+
+Even though Thanos comes with an increased architectural and operational complexity, we have to say after running
```

Comment:
after running it for a while, we think it is totally worth it

Reply:
Done